### DISCUSSION TITLE
LVN241HV4 and PREMTB100 replacement with microcontroller

### DISCUSSION THREAD

**Discussion by qurpi** [JanM321/esphome-lg-controller#86] [Show and tell]
First of all, thank you to JanM321 and Florian Brede for putting this together. No joke, I literally say a small prayer for you both every time I use the AC through my phone and how much of a blessing it is.

For context, our LG LVN241HV4 was installed at the entrance of our rental one bedroom apartment with the PREMTB100 wall controller, also installed at the entrance. The issue was, we have a narrow hallway before our apartment starts and this hallway is isolated from the rest of our house. As a result, the HVAC was working off temperatures in the hallway (the PREMTB100 has a built in temperature and humidity sensor) that were not entirely reflective of the areas where we spent majority our time i.e. lounge, kitchen and bedroom.

![20241117_143428](https://github.com/user-attachments/assets/b30fbd34-c5f0-49fc-9544-5df869e17579)

We first set out by getting an SHTC3 temperature sensor off of Amazon and setting it up with an ESP12-E microcontroller and having it show up on HomeAssistant (running on a docker container on a Pi 4).

After this, I ordered the ESP32 ESPRESSIF microcontroller off of Amazon and went about flashing it with the instructions mentioned. I don't have much experience with ESPHome (I primarily use Tasmota), as a result I put the firmware installation instructions into ChatGPT and had it break it down for me with the code which I could just copy and paste and run. This went pretty smoothly, I used ChatGPT to generate an encryption key from me.

Once this was showing up on HomeAssistant, I went ahead and started the process of ordering the PCBs via JLCPCB. They were short on one of the capacitors and so I replaced it accordingly (PANASONIC EEEFC1C100AR 10uF 16V 45mA@120Hz ±20% SMD,D4xL5.4mm replaced with (ST(Semtech) CK1C100MCRC54)). I used ChatGPT to verify this replacement was appropriate.

Ordered 5 and they were delivered in about 6 business days. 

![20241116_142549](https://github.com/user-attachments/assets/919a594f-feff-4dac-b797-d478c3d0f3a8)

Then went about looking for the CN-REMO port. The LVN241HV4 indoor unit is fairly large. I closed the breakers to the unit and then went about opening these four screws. This revealed another panel inside with 2 screws that I opened which revealed the circuit board. I found the CN-REMO port but it had a wire plugged into it already. I traced it back and realized this was going to the PREMTB100 wall controller. 

![IMG-20241116-WA0000](https://github.com/user-attachments/assets/3b03f3f4-c4fb-4ce9-b81a-b95728546a96)

I decided on initially trying to replace the PREMTB100 wall controller altogether. This is what things looked like when I took apart the controller. I wired those these wires to a JST-XH plug and plugged everything in to the microcontroller/PCB unit. It immediately come online on HomeAssistant. I was able to get the AC working and could hear the fan coming on but I was uneasy that I had no way of verifying whether the correct modes or temperate was set.

I decided to try splicing the JST-XH wires on to the wires that fed the PREMTB100 wall controller and screw them both into the wall controller's screw ports. 

![20241117_143434](https://github.com/user-attachments/assets/b1d78010-59bd-4cdc-8072-d769f20b0d8b)

At this point, I could control the AC with both the LG PREMTB100 wall controller and HomeAssistant. Any changes I made on HomeAssistant would reflect on the LG PREMTB100 wall controller (after about a 10 - 15 sec delay). However, I noticed that sometimes, the PREMTB100 wall controller would override the HomeAssistant controls. I realized that the PREMTB100 was in "Master" mode and so was the ESP32 microcontroller. I dug up an LG installation manual for the PREMTB100 and then entered installer mode to change it to "Slave" mode.

![20241117_143417](https://github.com/user-attachments/assets/61b315fa-12ed-47fa-b5f1-74451e49c606)

After this, there were no conflicts and everything syncs up very nicely. For some reason, I can change the AC settings on the PREMTB100 and they reflect on HomeAssistant. I'm not sure why this works but its super convenient to have both options.

**Comment by JanM321** [comment]
> First of all, thank you to JanM321 and Florian Brede for putting this together. No joke, I literally say a small prayer for you both every time I use the AC through my phone and how much of a blessing it is.

You’re welcome! I’ve been using this for 1.5 years now and I’m still very happy this all worked out, especially the external temperature sensor support. Home Assistant automations are neat too for changing settings based on time of day etc.

Thanks for the detailed write-up! It’s always nice to see how this is being used.

> For some reason, I can change the AC settings on the PREMTB100 and they reflect on HomeAssistant. I'm not sure why this works but its super convenient to have both options.

Slave controllers can change the basic settings and the ESPHome controller accepts setting changes from both the HVAC unit and from slave controllers, so that’s working as expected :)

I did only minimal testing of this because I don’t use multiple controllers myself, so I’m glad it works well.

**Comment by Adam1986** [comment]
@qurpi, I've just come across your insightful post.

I'm hoping to venture down this route and do away with LG ThinQ Cloud Integration.

My brother has a setup of wired wall controllers and is also thinking of doing the same.  However, what did you do to hide the ESP hardware given you've located it alongside the wall controller.  My brother cannot get to the connector on the unit to "split" it there, meaning he would have to do something at the current wall controllers.

Also going by the pictures you attached, I take it the USB cable was just there for the purpose of setting up and was then disconnected once up and running? 

Any further help/directions would be great.
